### PR TITLE
Integration ENS with IPFS

### PR DESCRIPTION
**Pull request by PhyrexTsai**
## Simple Summary
This implementation is to build a way to resolves ENS domains (.eth) to IPFS hash and redirects to url content.

## Motivation
MetaMask browser extension is great project helps users to interact with Ethereum, and furthermore can be more powerful by add content resolve by ENS.

## Specification
Resolve from ENS Public Resolver `getContent` hash to IPFS hash, and redirect url to `ipfs.infura.io`.

## Video
https://www.dropbox.com/s/8g8y42z6w7e6bcj/metamask.mov?dl=0

### CODE REVIEW

**Comment by kumavis** [comment]
@PhyrexTsai this is awesome, please give us time to review it!

File: app/404.html [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    }
+    h2 > a{
+        color: #1b243d;
+    }
+  </style>
+</head>
+<body>
+    <div class="app">
+        <img src="./images/404.png" alt="">
+        <h2>Powered by <a href="https://www.portal.network/">Portal Network</a></h2>
```
Comment (danfinlay) [inline app/404.html]: I wasn't sure how I felt about this at first, but you know, getting credit for what you do seems pretty fair. I'm fine leaving it up, at least for a while. The same way we link to 409H for the phishing detector.

File: app/manifest.json [hunk 2]
```diff
@@ -59,14 +59,19 @@
...
   ],
   "externally_connectable": {
     "matches": [
       "https://metamask.io/*"
     ]
-  }
-}
\ No newline at end of file
+  },
+  "content_security_policy": "script-src 'self' 'unsafe-eval'; object-src 'self'"
```
Comment (danfinlay) [inline app/manifest.json]: Is there a way to not include the `unsafe-eval` permission?

File: app/manifest.json [hunk 3]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "__MSG_appName__",
   "short_name": "__MSG_appName__",
-  "version": "4.6.0",
+  "version": "4.6.2",
```
Comment (danfinlay) [inline app/manifest.json]: Please leave the version bumping to us.

File: app/manifest.json [hunk 4]
```diff
@@ -59,14 +59,19 @@
...
-    "https://*.infura.io/"
+    "https://*.infura.io/",
+    "activeTab",
+    "webRequest",
+    "*://*.eth/"
   ],
   "web_accessible_resources": [
-    "inpage.js"
+    "inpage.js",
+    "*"
```
Comment (danfinlay) [inline app/manifest.json]: I don't think we want to make *all* resources web-accessible. This seems like a potential privacy breech.

I think you should specifically list the files needed to display the page statuses (html, images).

File: app/scripts/lib/ipfsContent.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        extension.tabs.getSelected(null, tab => {
+            extension.tabs.update(tab.id, { url: 'loading.html' })
+
+            clearTime = setTimeout(() => {
+                return extension.tabs.update(tab.id, { url: '404.html' })
+            }, 60000)
+
+            resolver.resolve(name, provider).then(ipfsHash => {
+                clearTimeout(clearTime)
+                let url = 'https://gateway.ipfs.io/ipfs/' + ipfsHash
```
Comment (danfinlay) [inline app/scripts/lib/ipfsContent.js]: Are we sure we want to hit IPFS gateway and not Infura?

Sending this comment to Infura for comment.

File: app/scripts/lib/resolver.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    } else {
+      return reject('unsupport')
+    }
+  })
+}
+
+function getProvider (type) {
+  switch (type) {
+    case 'mainnet':
+      return 'https://mainnet.infura.io/'
```
Comment (danfinlay) [inline app/scripts/lib/resolver.js]: I would rather we use the user's provider, provided by MetaMask, to reduce our number of blockchain connections, but for this feature, I can see the benefit of this simplicity, there aren't other chains w/ the ENS resolver anyway, this _could_ be optimized later...

**Review by danfinlay — COMMENTED** [review]
Just commenting for now, there are a few other opinions I'd like to get on parts of this.

**Comment by egalano** [comment]
@danfinlay you can use https://ipfs.infura.io for this. We've put a lot of work into preparing for additional traffic for IPFS.

File: app/scripts/lib/ipfsContent.js (hunk 5 above)
Comment (PhyrexTsai) [reply app/scripts/lib/ipfsContent.js]: Sure, I have changed to the `ipfs.infura.io`

**Comment by danfinlay** [comment]
@PhyrexTsai this would be very nicely accompanied by:

- A tutorial of how to post sites in this way.
- Some sample sites that work with this. The gif goes by quickly, we want to be able to show it off in full force when it comes out.
- A blog post for when we publish. I'll draft this one, but offer you a review before I post it.

**Comment by danfinlay** [comment]
I've verified this works with `phyrextsai.eth`, it looks good, sorry for taking so long to try.

When I got it working, at first it usually didn't work, so I think it maybe only works with new tabs after installation, not sure why it was intermittently not working on first uses.

I mostly just want a guide on creating compatible sites, so others can build on this as soon as we release.

**Review by danfinlay — DISMISSED** [review]
Works alright, has some issues with first time usage, and it doesn't save user's typed items, but does work with links like http://phyrextsai.eth, so that's cool, a blog post will really make this valuable to the community, but that doesn't need to block inclusion.

**Comment by PhyrexTsai** [comment]
@danfinlay Sorry for the late reply, I have write a tutorial of how to post websites by using IPFS and interact with ENS.

Tutorial link: [https://gist.github.com/PhyrexTsai/cffcbfa1d752b9cf817d920dfcd1ec9f](https://gist.github.com/PhyrexTsai/cffcbfa1d752b9cf817d920dfcd1ec9f)

And also have some website already deployed
- http://phyrextsai.eth
- http://christophershen.eth
- http://hozt.portalnetworkweb.eth
- http://teamnull.bluzellehackathon.eth
- http://cosmos.bluzellehackathon.eth
- http://dturker.bluzellehackathon.eth

File: app/scripts/background.js [hunk 7]
```diff
@@ -154,6 +158,7 @@ async function initialize () {
   const initLangCode = await getFirstPreferredLangCode()
   await setupController(initState, initLangCode)
   log.debug('MetaMask initialization complete.')
```
Comment (PhyrexTsai) [inline app/scripts/background.js]: For the first time can not connect with, the issue that the `initState.NetworkController.provider` is `undefined`, so I think we can have a default value for this issue.

The fix code look likes:
```javascript
let defaultProvider = initState.NetworkController === undefined ? { type: "mainnet" } : initState.NetworkController.provider;
ipfsContent(defaultProvider);
```

**Comment by danfinlay** [comment]
The guide looks great @PhyrexTsai. Would you mind if we reposted that on our blog when it's time to publish?

**Comment by PhyrexTsai** [comment]
@danfinlay Sure, that will be great, thank you very much

**Comment by kumavis** [comment]
we're cutting a bug fix release tomorrow, then this will go in after with the next batch of larger features 😸
thank you for your patience

**Comment by Arachnid** [comment]
Any way to make this resolve to the IPFS content, rather than redirecting to it?

**Comment by pimato** [comment]
@PhyrexTsai  @danfinlay @kumavis thanks for creating/adding this very important feature. Can´t wait to move the oasis products to IPFS/ENS.

**Comment by alexvandesande** [comment]
@Arachnid does a plugin has the right to change DNS hosts and actually register a "eth" TLD, so it looks like a traditional address? Honestly it would be great, but I would understand if google didn't permitted it as it would make phishing powers of extensions much worse

**Comment by bgits** [comment]
Shouldn't this feature use IPLD instead of IPFS? The `content` field could be a hash pointing to content stored in all kinds of protocols not just IPFS.

**Comment by Georgi87** [comment]
Are pathnames forwarded? E.g. if I go to mywebsite.eth/home it should redirect to <ipfs-redirect>/home

Event better would be to resolve IPFS content directly. Is this possible to do in a chrome extension @PhyrexTsai ?

**Comment by decanus** [comment]
From what I’ve seen, you can resolve content directly if you register a new scheme. So like ipfs:// instead of http://

**Comment by PhyrexTsai** [comment]
@Georgi87 I have committed the feature to let the mechanism in the pathnames of .eth domain working as the same in the ordinary domain.

**Comment by danfinlay** [comment]
A few files are not passing our linting rules. We can either:

- Fix these on this branch.
- Add these files to the lint exceptions list.
- Move these files outside of this repository.

The affected files are:

- `app/scripts/lib/ipfsContent.js`
- `app/scripts/lib/contracts/registrar.js`

These errors can be seen here:
https://circleci.com/gh/MetaMask/metamask-extension/31652?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Or by running: `npm run lint`.

These may be automatically fixable by running `npm run lint:fix`.

**Comment by brunobar79** [comment]
I'm gonna submit a PR to this branch fixing the linter error

**Comment by PhyrexTsai** [comment]
@danfinlay @brunobar79 Thanks for helping, I've merged the PR on this branch.

**Review by danfinlay — APPROVED** [review]
Let's do this!

**Comment by jdetychey** [comment]
can't wait for this :rocket:

**Comment by Arachnid** [comment]
Note that not all ENS names will end in .eth. Can you make this work for other TLDs too?

**Comment by bgits** [comment]
@Arachnid What are the planned TLD's or how will new TLD's be proposed and added?

**Comment by Arachnid** [comment]
@bgits .luxe and .dds are likely to be  added in the very near future, and correspond to their DNS equivalents. In the longer run, nearly any DNS TLD will be supported via DNSSEC proofs.

File: app/manifest.json [hunk 8]
```diff
@@ -59,7 +59,10 @@
     "unlimitedStorage",
     "clipboardWrite",
     "http://localhost:8545/",
-    "https://*.infura.io/"
+    "https://*.infura.io/",
+    "activeTab",
+    "webRequest",
+    "*://*.eth/"
```
Comment (mryellow) [inline app/manifest.json]: This seems to appear in permissions dialog for Chrome as "eth websites" without an indication it's talking about a TLD. Leaving the user wondering "What is an eth website?"

**Pull request merged**